src/parser.rs

```diff
@@ -1,7 +1,45 @@
 use std::any::Any;
 
+#[derive(Clone, Debug)]
+pub struct Input<'a> {
+    cursor: usize,
+    value: &'a str,
+}
+
+impl<'a> Input<'a> {
+    pub fn new(input: &'a str) -> Self {
+        Input {
+            cursor: 0,
+            value: input,
+        }
+    }
+
+    pub fn head(&mut self, pattern: &str) -> &'a str {
```

Comment:
This should return an `Option`, not an empty string, when it can't find the pattern

Reply:
The function only returns an empty string if A) the end of the string has been reached B) the pattern is matched right away. The function returns the head up until the patter, everything after the pattern is the tail.

Reply:
All right, that's fine, then. It seems more idiomatic to me to return an `Option`, but it's a matter of opinion.

Reply:
Currently head is a weird function since it advances the cursor for you.. it does not operate as regular head and tail as you know it from functional languages.